File: cpp/3612.process-string-with-special-operations-i.cpp
#include <algorithm>
#include <string>

using std::string;

class Solution {
public:
    string processStr(string s) {
        string res;
        // # - duplicate
        // % - reverse
        // * - pop last if exists

        for (char ch : s) {
            if (ch == '#') {
                res.append(res);
            } else if (ch == '%') {
                std::reverse(res.begin(), res.end());
            } else if (ch == '*') {
                if (res.size()) {
                    res.pop_back();
                }
            } else {
                res.push_back(ch);
            }
        }

        return res;
    }
};

File: rust/3612.process-string-with-special-operations-i.rs
impl Solution {
    pub fn process_str(s: String) -> String {
        let mut res = Vec::new();

        for ch in s.chars() {
            match ch {
                '*' => {
                    res.pop();
                }
                '#' => {
                    res.extend_from_within(..);
                }
                '%' => {
                    res.reverse();
                }
                _ => {
                    res.push(ch);
                }
            }
        }

        res.into_iter().collect()
    }
}
